README


1. DESCRIZIONE

Programmino a linea di comando che splitta/codifica i file audio rippati 
su una sola traccia (dotati di cue-file) per preservare la fedeltà al CD 
originale. Questo tipo di tecnica viene definita "cue splitting".
Pysplitcue consente di splittare e allo stesso tempo decodificare/codificare 
traccie in formato WAV, FLAC e APE. La funzione che svolge il programma 
necessita della presenza del file *.cue nella medesima directory della 
traccia musicale.
I file di origine non saranno rimossi ne modificati in alcun modo.

combinazioni di conversione:

Splitting da .wav a .wav, da .wav a .flac, da .wav ad .ape  
da .flac a .wav, da flac a flac, da .flac ad .ape,
da .ape a .wav, da .ape a .flac, da ape ad ape


Dipendenze:
*  python2.6
*  flac
*  cuetools
*  shntool
*  mac

L'installazione comprende una pagina man e un help (-help o -h)


CHANGELOG
+--------------------------------+
January  06 2022  Version 2.1.5
+--------------------------------+

- Added support for mp3, ogg output formats.
- auto-detect CUE shee file encoding.
- Rewrite the tests with the unittest module.
- Now pysplitcue can also be used interactively as a Python module.
- Improved exception handling.


+--------------------------------+
December 28 2021  Version 2.1.4
+--------------------------------+

- Fixed issue using `-o` option for recursive directory creation.
- Code refactoring
- Made unittest test cases

+--------------------------------+
December 26 2021  Version 2.1.3
+--------------------------------+

- Fixed `ask` option when there is nothing to overwriting #8.
- Fixed 'WARNING: File already exists:' if you save tracks in another folder
  and tracks already exists on cue sheet path.
- Added the ability to include relative or absolute paths.
- Fixes minor bugs


+--------------------------------+
December 25 2021  Version 2.1.2
+--------------------------------+

- Changed the source code launcher name to pysplitcue.py .
- Changed package name to pysplitcue
- Changed splc.py module to pysplitcue.py.
- Fixed "No executable after pip install" #6
- Changed `-p`, `--preferred-format` option to `-f`, `format-type`
- Updated man page, help and README.md.
- Updated actions for parsing arguments.
- Fixed COPYING in MANIFEST.in file.
- Added new option to set actions to prevent or allow overwriting of
  existing files.
- Improved console messages (made some cosmetic text).

+--------------------------------+
December 15 2021  Version 2.1.1
+--------------------------------+

- Changed, input file must be CUE sheet now.
- Fixed which(cuetag.sh) variable on splc.py .
- Changed `subprocess.check_call` to `subprocess.run`
- Added option to move audio tracks to an output folder of your choice.
- Fixed `warning: number of files does not match number of tracks` see #4.
- Use a temp dir to fix #4 .
- Rewrite man page and help.
- The flac format is the default in the choice of output formats.

+--------------------------------+
December 07 2021  Version 2.0.0
+--------------------------------+

- Fixed missing quotes in the filenames #2
- Some code refactoring.
- Improved description and examples
- Improved man page
- Fixed setup.py

+--------------------------------+
August 8 2019  Version 1.0.0
+--------------------------------+

- Support guaranteed only on Python3
- Add checking dependencies
- Add argparser integrity
- Checking the input file for determine supported audio
  formats and cue sheet presence.

+--------------------------------+
Novembar 24 2017  Version 0.6.3
+--------------------------------+

- Improved error management, now it's a little better commented

- I have separated the splitting process and the tagging process
  for improve errors management


+--------------------------------+
Novembar 21 2017  Version 0.6.2
+--------------------------------+

- Add a error management interface for shell output

- Improved python code somewhere

- Resolved the cuetag binaries name on slackware 14.2


+--------------------------------+
January 26 2015  Version 0.6.1
+--------------------------------+

- This is first stable release

- Translation done completely in English

- Make a setup.py and MANIFEST.in for building a package for your Linux
  distribution (Debian and derived, Slackware)

- It was clean the main executable code

+--------------------------------+
August 25 2012  Version 0.6
+--------------------------------+
- Modified the program manual (I hope it is more understandable)

- Adjusted positional parameters in this order:
  program_name > option > path-name (optional)


+--------------------------+
Nov. 09  2010  Version 0.5
+--------------------------+

Initial not public release


pysplc/splc.py
"""
First release: 25/08/2012

Name: pysplitcue
Porpose: wraps the shnsplit and cuetag commands
Platform: Mac OsX, Gnu/Linux
Writer: jeanslack <jeanlucperni@gmail.com>
license: GPL3
Rev: Dec 15 2021
Code checker: flake8 and pylint
####################################################################

This file is part of pysplitcue.

    pysplitcue is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    pysplitcue is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with pysplitcue.  If not, see <http://www.gnu.org/licenses/>.
"""
import sys
import os
from shutil import which
import shutil
import argparse
import subprocess
import tempfile
from pysplc.str_utils import information

# strings of information
INFO = information()
DATA = INFO[0]


def check_executables():
    """
    check for binaries: Returns the cuetag script
    (On Slackware 14.1 it is named cuetag.sh)
    """
    if which('cuetag'):
        cuetag = which('cuetag')
    elif which('cuetag.sh'):
        cuetag = which('cuetag.sh')
    else:
        sys.stderr.write("\033[31;1mERROR:\033[0m pysplitcue: 'cuetag' "
                         "is required, please install 'cuetools'.\n")
        sys.exit(1)

    if not which('shntool'):
        sys.stderr.write("\033[31;1mERROR:\033[0m pysplitcue: 'shntool' "
                         "is required, please install it.\n")
        sys.exit(1)

    return cuetag
# ----------------------------------------------------------#


def dependencies():
    """
    Check for dependencies
    """
    listing = ['flac', 'mac', 'wavpack', 'shntool', 'cuebreakpoints',
               'cueconvert', 'cueprint']
    for required in listing:
        # if which(required):
        if which(required, mode=os.F_OK | os.X_OK, path=None):

            print(f"Check for: '{required}' ..Ok")
        else:
            print(f"Check for: '{required}' ..Not Installed")

    if which('cuetag', mode=os.F_OK | os.X_OK, path=None):
        print("Check for: 'cuetag' ..Ok")

    elif which('cuetag.sh', mode=os.F_OK | os.X_OK, path=None):
        print("Check for: 'cuetag.sh' ..Ok")
    else:
        print("Check for: 'cuetag' ..Not Installed")
# ----------------------------------------------------------#


def run_process_splitting(filecue, audiofile, preferred_format, tmpdir):
    """
    Compare the dict keys to get the matched string value
    required by `shnsplit` command.
    """
    name = os.path.splitext(audiofile)[0]
    inputformat = os.path.splitext(audiofile)[1].replace('.', '', 1)

    cmd_split = {'wav:wav': (f'shnsplit -o wav -f "{filecue}" '
                             f'-t "%n - %t.split" -d "{tmpdir}" "{name}.wav"'),

                 'wav:flac': (f'shnsplit -o flac -f "{filecue}" '
                              f'-t "%n - %t" -d "{tmpdir}" "{name}.wav"'),

                 'wav:ape': (f'shnsplit -o ape -f "{filecue}" '
                             f'-t "%n - %t" -d "{tmpdir}" "{name}.wav"'),

                 'flac:wav': (f'shnsplit -o wav -f "{filecue}" '
                              f'-t "%n - %t" -d "{tmpdir}" "{name}.flac"'),

                 'flac:flac': (f'shnsplit -o flac -f "{filecue}" '
                               f'-t "%n - %t" -d "{tmpdir}" "{name}.flac"'),

                 'flac:ape': (f'shnsplit -o ape -f "{filecue}" '
                              f'-t "%n - %t" -d "{tmpdir}" "{name}.flac"'),

                 'ape:wav': (f'shnsplit -o wav -f "{filecue}" '
                             f'-t "%n - %t" -d "{tmpdir}" "{name}.ape"'),

                 'ape:flac': (f'shnsplit -o flac -f "{filecue}" '
                              f'-t "%n - %t" -d "{tmpdir}" "{name}.ape"'),

                 'ape:ape': (f'shnsplit -o ape -f "{filecue}" '
                             f'-t "%n - %t.split" -d "{tmpdir}" "{name}.ape"')
                 }

    if f'{inputformat}:{preferred_format}' in cmd_split:
        print("\nOn splitting audio tracks...")
        split = cmd_split[f'{inputformat}:{preferred_format}']
        try:
            subprocess.run(split, check=True, shell=True)

        except subprocess.CalledProcessError as err:
            sys.exit(f"\033[31;1mERROR:\033[0m {err}")

        else:
            print("\033[1m...done splitting\033[0m")
    else:
        msg = f"Unsupported input file format '{inputformat}'"
        sys.exit(f"\033[31;1mERROR:\033[0m {msg}")
# ----------------------------------------------------------#


def run_process_tagging(filename, audiofile, preferred_format):
    """
    Compare the dict keys to get the matched string value
    required by `cuetag` command.
    """
    exitstatus = None
    cuetag = check_executables()
    inputformat = os.path.splitext(audiofile)[1].replace('.', '', 1)

    cmd_tag = {'wav:flac': f'{cuetag} "{filename}" *.flac',
               'flac:flac': f'{cuetag} "{filename}" *.flac',
               'ape:flac': f'{cuetag} "{filename}" *.flac'}

    if f'{inputformat}:{preferred_format}' in cmd_tag:
        print("\nApply tags on audio tracks...")
        tag = cmd_tag[f'{inputformat}:{preferred_format}']

        try:
            subprocess.run(tag, check=True, shell=True)

        except subprocess.CalledProcessError as err:
            sys.exit(f"\033[31;1mERROR:\033[0m {err}")

        else:
            print("\033[1m...done tagging\033[0m")
    else:
        exitstatus = (f"Unsupported file format for "
                      f"tagging '{preferred_format}' ..skip")

    return exitstatus
# ----------------------------------------------------------#


def make_subdir(outputdir, tmpdir):
    """
    By giving the -o option, the user can supply a path
    with the name of a given folder for the output files.
    """
    if not outputdir == '.':
        try:
            os.mkdir(outputdir, mode=0o777)
        except OSError:
            print("\033[32;1mINFO:\033[0m A destination folder "
                  "already exists ..skip")

    for track in os.listdir(tmpdir):
        try:
            shutil.move(os.path.join(tmpdir, track),
                        os.path.join(outputdir, track))
        # Catching too general exception Exception (fixme)
        except Exception as err:
            sys.exit(f"\033[31;1mERROR:\033[0m {err}")
# ----------------------------------------------------------#


def cuefile_reading(fname, suffix):
    """
    Given a *.cue sheet file, it extracts the titles
    of the audio tracks to be split. Returns a tuple
    with the relative name of the audio file and a
    dictionary containing the names of the audio tracks .
    """
    num = 'TITLE'
    titletracks = {}

    with open(fname, 'r', encoding='utf8') as cue:
        filecue = cue.readlines()

    for line in filecue:
        if 'FILE' in line:
            titletracks['FILE'] = line.split('"')[1]

        if 'TRACK' in line:
            num = line.strip().split()[1]

        if 'TITLE' in line:
            title = line.split('"')[1]
            titletracks[num] = (f'{num} - {title}.{suffix}')

    return titletracks
# ----------------------------------------------------------#


def cuefile_check(filename):
    """
    Checks the CUE file. Returns True if error, None otherwise
    """
    error = None

    if not os.path.isfile(filename):
        error = True
    elif os.path.splitext(filename)[1] not in ('.cue', '.CUE'):
        error = True

    return error
# ----------------------------------------------------------#


def main():
    """
    Parser of the users inputs (evaluates positional arguments)
    using the argparser module.
    """
    parser = argparse.ArgumentParser(
                prog=DATA['prg_name'],
                description=DATA['short_decript'],
                # add_help=False,
                )
    parser.add_argument(
                '-v', '--version',
                help="Show the current version and exit",
                # action="store_true",
                action='version',
                version=f"pysplitcue v{DATA['version']} - {DATA['release']}",
                       )
    parser.add_argument(
                '-c', '--check-requires',
                help="List of installed or missing dependencies",
                action="store_true",
                required=False,
                       )
    parser.add_argument(
                '-p', '--preferred-format',
                choices=["wav", "flac", "ape"],
                help="Preferred audio format to output, default is flac",
                required=False,
                default='flac',
                        )
    parser.add_argument(
                '-i', '--input-cuefile',
                metavar='IMPUTFILE',
                # type=str,
                help=("INPUTFILE must be a CUE sheet with '.cue' "
                      "filename extension"),
                # dest='enable_config',
                # action='store_true',
                # nargs='?',
                # default=PWD,
                action="store",
                required=True,
                )
    parser.add_argument("-o", "--output-dir",
                        action="store",
                        type=str,
                        dest="outputdir",
                        help="Output directory, default '.'",
                        required=False,
                        default='.')

    args = parser.parse_args()

    if args.check_requires:
        dependencies()

    elif args.input_cuefile:
        filename = args.input_cuefile
        dirname = os.path.dirname(filename)
        fname = os.path.basename(filename)
        outputdir = args.outputdir
        suffix = args.preferred_format

        error = cuefile_check(filename)
        if error:
            parser.error(f"Invalid file: '{filename}'\n"
                         f"Provide a CUE sheet file (*.cue or *.CUE format).")
        else:
            os.chdir(dirname)
            titletracks = cuefile_reading(fname, suffix)
        if not titletracks:
            sys.exit(f"pysplitcue: error: Unable to read: '{filename}'")
        else:
            with tempfile.TemporaryDirectory(suffix=None,
                                             prefix='pysplitcue_',
                                             dir=None) as tmpdir:
                run_process_splitting(fname,
                                      titletracks['FILE'],
                                      suffix,
                                      tmpdir
                                      )
                os.chdir(tmpdir)
                runtag = run_process_tagging(filename,
                                             titletracks['FILE'],
                                             suffix
                                             )
                if runtag:
                    print(f"\033[33;1mWARNING:\033[0m {runtag}")

                os.chdir(dirname)
                make_subdir(outputdir, tmpdir)

            print("\033[1mFinished!\n\033[0m")
            sys.exit(0)


if __name__ == '__main__':
    main()
